src/server.h

```diff
@@ -1235,7 +1236,6 @@ typedef struct client {
     dictEntry *cur_script;  /* Cached pointer to the dictEntry of the script being executed. */
     time_t lastinteraction; /* Time of the last interaction, used for timeout */
     time_t obuf_soft_limit_reached_time;
-    int authenticated;                   /* Needed when the default user requires auth. */
```

Comment:
The change LGTM. 

on a second thought though, I wonder if we should use bitfields instead and the code will be a lot more compact as a result and more readable/debuggable.

Reply:
This mainly focuses on optimizing memory and improving performance. Debuggable can be optimized at the tool level, such as GDB functions.

Reply:
This proposal feels great, currently it seems that the flag is already quite large. Even not just the client. Just like servers also have many such fields. Alternatively, a separate topic can be initiated to review the benefits of this optimization.

Reply:
> on a second thought though, I wonder if we should use bitfields instead and the code will be a lot more compact as a result and more readable/debuggable.

My only counter point is that we basically already have 51 other bitfields on the client, I'm not sure it makes sense to start moving the flags to bitfields now since we don't get any of the benefit.

Reply:
There is another more minor downside I suppose, which is that we can more easily set a lot of bits with this structure. We do a "reset" client at one point where we update the entire flag structure.

Reply:
Maybe we could do something like:
```
typedef struct client {
    union {
        uint64_t flags; /* Client flags: CLIENT_* macros. */
        struct {
            uint64_t not_addressable : 52;
            uint64_t authenticated : 1;
            uint64_t reserved : 11;
        };
    };
...
}
```

Best of both world perhaps? We can still flush the entire flags with a single command but each flag is still addressable.

Reply:
LGTM, Union is a great solution that makes it more convenient for us to complete the renovation.

Reply:
I like it :)